apis/scheduling/v1alpha1/reservation_types.go

```diff
@@ -57,6 +51,17 @@ type ReservationSpec struct {
 	// `expires` and `ttl` are mutually exclusive. Defaults to being set dynamically at runtime based on the `ttl`.
 	// +optional
 	Expires *metav1.Time `json:"expires,omitempty"`
+	// By default, the resources requirements of reservation (specified in `template.spec`) is filtered by whether the
+	// node has sufficient free resources (i.e. Reservation Request <  Node Free).
+	// When `preAllocation` is set, the scheduler will skip this validation and allow overcommitment. The scheduled
+	// reservation would be waiting to be available until free resources are sufficient.
+	// +optional
+	PreAllocation bool `json:"preAllocation,omitempty"`
+	// By default, reserved resources are always allocatable as long as the reservation phase is Available. When
+	// `AllocateOnce` is set, the reserved resources are only available for the first owner who allocates successfully
+	// and are not allocatable to other owners anymore.
+	// +optional
+	AllocateOnce bool `json:"allocateOnce,omitempty"`
```

Comment:
Alternative:

- wrap `AllocateOnce` and `PreAllocation` into a new type `AllocationOption`:

```go
type AllocationOption struct {
    PreAllocation bool
    Once bool
}
```

Reply:
How to deal with the concurrency problem of bind and changing reservation state?

Reply:
Currently, the reservation plugin simply uses optimistic concurrency control for these cases.

e.g. When multiple pods allocate(Bind) one reservation almost simultaneously, the plugin

1. gets the latest version of the reservation to bind
2. double-checks if the reservation still matches the pod
3. updates the reservation's status

If a reservation is allocated and becomes `Succeeded`, other pods which match the old version will fail in step 2.

This may introduce some updating conflicts. We can further reduce the probability of concurrency conflicts by the following:

- Only the scheduler which specified by the reserve pod (via `spec.schedulerName`) can change the phase of the reservation.
- The scheduling pod only can allocate the available reservations which specify the same scheduler.

Reply:
The update of reservation and the bind phase of the pod cannot maintain atomicity, thus the semantics of one cannot be guaranteed?

et. Pod A use reservation B, bind Pod A succeed but update reservation B failed, and after scheduler restart, reservation B will continue to exist.

Reply:
> The update of reservation and the bind phase of the pod cannot maintain atomicity, thus the semantics of one cannot be guaranteed?
> 
> et. Pod A use reservation B, bind Pod A succeed but update reservation B failed, and after scheduler restart, reservation B will continue to exist.

In the current design, the Pod A can succeed the binding only if it has updated reservation B successfully. Otherwise, it would fail the binding cycle.